### PR TITLE
Refactor notification controller spec to notification request spec

### PR DESCRIPTION
**Pull request by Ashwithabg**
#1831

### CODE REVIEW

File: spec/requests/notifications_spec.rb [hunk 1]
```diff
@@ -1,24 +1,23 @@
 # frozen_string_literal: true
 
-RSpec.describe NotificationsController, type: :controller do
+describe 'Notifications', type: :request do
```
Comment (bennpham) [inline spec/requests/notifications_spec.rb:3]:
```suggestion
describe 'Notification', type: :request do
```

Nitpick: I noticed that some of the request _spec are plural and some are singular. I think for consistency, let's stick to singular to match the rest of the example request specs: https://relishapp.com/rspec/rspec-rails/docs/request-specs/request-spec . Same goes for filename.

**Review by bennpham — COMMENTED** [review]
I'll probably get through reviewing the rest of the tests later, but I'm finding the multiple context blocks a bit tricky to follow. It's DRY, but not DAMP so it's hard for me to follow which action I'm trying to assert as I go down the tests. 

I think multiple nested context with `and` is an anti pattern. There are some projects I work on with default rubocop rule to prevent more than 2-3 nested context/describe blocks. I'd change some of them to say `context 'when ...' do` and merge some of those together.

File: spec/requests/notifications_spec.rb (hunk 1 above)
Comment (tyler-wel) [reply spec/requests/notifications_spec.rb:3]: I've always done it to match the controllers/routing. So if it's `notifications_controller` or `GET /notifications`, I'd do `notifications_spec`. But there's also some inconsistent naming in the controllers too so 🤷‍♀️

File: spec/requests/notifications_spec.rb [hunk 2]
```diff
@@ -1,24 +1,23 @@
...
       before do
-        allow(controller).to receive(:current_user).and_return(user)
-        request.env['HTTP_REFERER'] = previous_page
+        sign_in user
 
-        delete :destroy, format: format, params: { id: notification_id }
+        delete notification_path(id: notification_id), headers: headers
       end
 
       context 'and the notification to be deleted exists' do
```
Comment (tyler-wel) [inline spec/requests/notifications_spec.rb:23]: These names should be refactored. "and" naming should not be used.
https://www.betterspecs.org/#contexts

File: spec/requests/notifications_spec.rb [hunk 3]
```diff
@@ -28,15 +27,18 @@
...
               expect(response).to redirect_to previous_page
             end
           end
 
           context 'and the requested format is json' do
-            let(:format) { 'json' }
+            let(:headers) do
+              { 'HTTP_REFERER' => previous_page,
+                'ACCEPT' => 'application/json' }
+            end
```
Comment (tyler-wel) [inline spec/requests/notifications_spec.rb:41]: I'm personally not a fan of this code format
```suggestion
            let(:headers) do
              { 
                'HTTP_REFERER' => previous_page,
                'ACCEPT' => 'application/json' 
              }
            end
```

File: spec/requests/notifications_spec.rb [hunk 4]
```diff
@@ -52,15 +54,18 @@
...
+            let(:headers) { { 'HTTP_REFERER' => previous_page } }
 
             it 'redirects the user back to where they were before' do
               expect(response).to redirect_to previous_page
             end
           end
 
           context 'and the requested format is json' do
-            let(:format) { 'json' }
+            let(:headers) do
```
Comment (tyler-wel) [inline spec/requests/notifications_spec.rb:65]: I feel like this headers let block is repeated way too many times. Could be extracted to a higher context and shared. 
Make the accept block `'ACCEPT' => format` and define `let(:format)` where needed

File: spec/requests/notifications_spec.rb [hunk 5]
```diff
@@ -124,7 +132,7 @@
 
     context 'when the user is signed in' do
       before do
-        allow(controller).to receive(:current_user).and_return(user)
+        sign_in user
```
Comment (tyler-wel) [inline spec/requests/notifications_spec.rb:135]: Don't need `do...end` here

```suggestion
      before { sign_in user }
```

**Review by tyler-wel — CHANGES_REQUESTED** [review]

File: spec/requests/notifications_spec.rb (hunk 1 above)
Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb]: Have updated name of the spec to singular.

File: spec/requests/notifications_spec.rb (hunk 2 above)
Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb]: Changed `and the` to `when`  in context names.

File: spec/requests/notifications_spec.rb (hunk 3 above)
Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb]: Formatted the headers.

File: spec/requests/notifications_spec.rb (hunk 5 above)
Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb]: Updated this

File: spec/requests/notifications_spec.rb (hunk 4 above)
Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb:63]: In few context blocks, we need headers to have JSON format. And for a few HTML. That's the reason why headers are duplicated.

Comment (tyler-wel) [reply spec/requests/notifications_spec.rb:63]: My recommendation is to reduce the amount of duplicated long blocks of code by moving the headers block to a higher context and defining the format where needed
```
context 'when user is signed in' do
  let(:headers) do
    { 
      'HTTP_REFERER' => previous_page,
      'ACCEPT' => format
    }
  end

  context '...' do
    let(:format) { 'application/json' }
  end

  context '...' do
    let(:format) { 'text/html' }
  end
end
```

Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb:63]: okay thank you. Will do this change.

Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb]: I have pushed above changes. Could you please review.

File: spec/requests/notifications_spec.rb [hunk 6]
```diff
@@ -69,19 +71,17 @@
...
       end
 
-      context 'and the notification to be deleted does not exist' do
+      context 'when notification to be deleted does not exist' do
         let(:notification_id) { 'something-fake' }
 
-        context 'and the requested format is html' do
-          let(:format) { 'html' }
-
+        context 'when requested format is html' do
```
Comment (tyler-wel) [inline spec/requests/notifications_spec.rb:77]: you should specify for all html contexts
```ruby
let(:format) { 'text/html' }
```

this ensure the expected behavior is actually happening

**Review by tyler-wel — CHANGES_REQUESTED** [review]
@bennpham you good?

**Review by bennpham — APPROVED** [review]
@tyler-wel If all of your other reviews look good. I'm overall fine with this. Just a few nitpicks like some do/end could be reduce to { } to save some line spaces but nothing seem logically off for me now when I'm going through 👍 👍

File: spec/requests/notifications_spec.rb [hunk 7]
```diff
@@ -1,66 +1,68 @@
...
   describe '#destroy' do
-    let(:user) { FactoryBot.create(:user1) }
-    let(:other_user) { FactoryBot.create(:user2) }
+    let(:user) { create(:user1) }
+    let(:other_user) { create(:user2) }
     let(:notification_owner) { user }
     let!(:notification) do
-      FactoryBot.create(:notification, user: notification_owner)
+      create(:notification, user: notification_owner)
     end
```
[the remark this answers is not in the document]
Comment (tyler-wel) [reply spec/requests/notifications_spec.rb:19]: Agreed, do..end here isn't necessary 👍

File: spec/requests/notifications_spec.rb (hunk 6 above)
Comment (Ashwithabg) [reply spec/requests/notifications_spec.rb:79]: have made the changes. Thanks for the feedback

**Review by tyler-wel — APPROVED** [review]

**Pull request merged**